### ISSUE TITLE
Docs improvement idea: explain claim mapping

### ISSUE THREAD

**Issue by astegmaier** [IdentityServer/IdentityServer4#1285]
I wanted to propose an enhancement to the docs based on my experience reading them to learn about IdentityServer4. I'm happy to work on a draft, but I wanted to run the idea by you guys first--let me know what you think!

The [docs on the UserInfo endpoint](https://github.com/IdentityServer/IdentityServer4/blob/release/docs/endpoints/userinfo.rst) contain the following statement:

> Depending on the granted scopes, the UserInfo endpoint will return the mapped claims (at least the openid scope is required).

I had set up a basic IdentityServer where my users were stored in a database managed through ASP.NET identity, was attempting to read the docs to figure out how to get my UserInfo endpoint to return claims beyond "sub" and "name." For example, I wanted to include "email" (which is built-in to IdentityUser) and a custom "birthday" claim (which I had added by creating an ApplicationUser class that extended IdentityUser). But these weren't getting included in "mapped claims," and it wasn't clear from the docs how I could fix this.

After lots of googling, randomly tried following the instructions on [this blog post](http://benfoster.io/blog/customising-claims-transformation-in-aspnet-core-identity) (which didnt' mention IdentityServer at all) to add a custom UserClaimsPrincipalFactory to my Services pipeline that creates an "email" and "birthdate" claim from User.Email and User.Birthday, and it seemed to do the trick.

```
public class AppClaimsPrincipalFactory : UserClaimsPrincipalFactory<ApplicationUser, IdentityRole>
    {
        public AppClaimsPrincipalFactory(
            UserManager<ApplicationUser> userManager,
            RoleManager<IdentityRole> roleManager,
            IOptions<IdentityOptions> optionsAccessor)
        : base(userManager, roleManager, optionsAccessor)
        { }

        public async override Task<ClaimsPrincipal> CreateAsync(ApplicationUser user)
        {
            var principal = await base.CreateAsync(user);


            if (!string.IsNullOrWhiteSpace(user.Email))
            {
                ((ClaimsIdentity)principal.Identity).AddClaims(new[] {
                    new Claim("email", user.Email),
                });
            }

            if (!string.IsNullOrWhiteSpace(user.Birthday.ToString()))
            {
                ((ClaimsIdentity)principal.Identity).AddClaims(new[] {
                    new Claim("birthdate", user.Birthday.ToString()),
                });
            }

            return principal;
        }
    }
```
...and adding this class to DI as follows:

`services.AddScoped<Microsoft.AspNetCore.Identity.IUserClaimsPrincipalFactory<ApplicationUser>, AppClaimsPrincipalFactory>();`

Two questions:

1. Is this the right way to accomplish my goals?
2. Would you be opening to a change to enhance the UserInfo documentation (or on a separate page) that unpacks what "mapped claims" are and how to control them? That way other people trying to do the same thing will find something helpful.

**Comment by Genbox** [comment]
If you add the IdentityServer4.AspNetIdentity package, there is a service extension called [AddIdentityServerUserClaimsPrincipalFactory](https://github.com/IdentityServer/IdentityServer4.AspNetIdentity/blob/29011a258b6e342cb6ea2909e9cead806906ba5e/src/IdentityServer4.AspNetIdentity/ServiceCollectionExtensions.cs) which adds a [UserClaimFactory](https://github.com/IdentityServer/IdentityServer4.AspNetIdentity/blob/29011a258b6e342cb6ea2909e9cead806906ba5e/src/IdentityServer4.AspNetIdentity/UserClaimsFactory.cs) to the pipeline.

By default it supports username, email and phone number, saved using JWT claim types.

**Comment by astegmaier** [comment]
@Genbox thanks for pointing that out! I tried it out by adding this line to my `ConfigureServices`:
```
services.AddIdentityServerUserClaimsPrincipalFactory<ApplicationUser, IdentityRole>();
```
That does seem to have the effect of adding the `preferred_username`, `email`, and `email_verified` fields to the UserInfo response. Interestingly, it doesn't seem to add `phone number`, but I don't know why.

I also stumbled across [this example](https://github.com/astegmaier/angular-auth-oidc-client-SSR-test/blob/master/src/IdentityServerWithIdentitySQLite/IdentityWithAdditionalClaimsProfileService.cs) of how to implement an `IProfileService` to accomplish a similar goal.

@brockallen and @leastprivilege - are any of these three methods supported? Are they documented anywhere? If not, would you like me to take a stab at explaining them in the context of trying to control the output of UserInfo in IdentityServer (I think I have a handle on the UserPrincipleClaimsFactory method, but I still have a lot of questions about `AddIdentityServerUserClaimsPrincipalFactory` and `iProfileService`, so I might not be able to do as good a job about those).

**Comment by brockallen** [comment]
I recently (and coincidentally) added docs for the profile service: http://docs.identityserver.io/en/release/reference/profileservice.html. How claims come back in the id_token vs. userinfo, IMO, are not really our responsibility to fully document, since those are spec concerns, and, IMO, consumers of IdentityServer are expected to know (at least generally) how the specs work.

Some of the other things you mention are ASP.NET Identity specific, and were also recently (and coincidentally) documented: http://docs.identityserver.io/en/release/reference/aspnet_identity.html. How the internals of ASP.NET Identity claim mapping, IMO, are not our responsibility to document. That's a Microsoft product and if they don't do it, then it's on them.

Having said all of the above, and given the docs I added related to these concepts I think I'm inclined to close this issue. If you feel that there is more to add to the docs I listed, then we'd love contributions via a PR. If you wanted to write up a separate blog post with more info and feel like doing the education on the specs and/or ASP.NET Identity, then feel free and we can link to it from our docs.
